--- a/net/minecraft/world/level/block/CropBlock.java
+++ b/net/minecraft/world/level/block/CropBlock.java
@@ -74,14 +_,35 @@
         return !this.isMaxAge(state);
     }
 
+    public static final it.unimi.dsi.fastutil.objects.Object2DoubleOpenHashMap<String> growthPercentIncrease = new it.unimi.dsi.fastutil.objects.Object2DoubleOpenHashMap<>();
+
     @Override
     protected void randomTick(BlockState state, ServerLevel level, BlockPos pos, RandomSource random) {
-        if (level.getRawBrightness(pos, 0) >= 9) {
+        if (level.getRawBrightness(pos, 0) >= 9 || org.bukkit.Bukkit.worldNamesWithAllowedGrowthWithLowBrightness.contains(level.getWorld().getName())) {
             int age = this.getAge(state);
             if (age < this.getMaxAge()) {
                 float growthSpeed = getGrowthSpeed(this, level, pos);
-                if (random.nextInt((int)(25.0F / growthSpeed) + 1) == 0) {
-                    level.setBlock(pos, this.getStateForAge(age + 1), Block.UPDATE_CLIENTS);
+                // Spigot start
+                double modifier = 100;
+                if (this == Blocks.BEETROOTS) {
+                    modifier = level.spigotConfig.beetrootModifier;
+                } else if (this == Blocks.CARROTS) {
+                    modifier = level.spigotConfig.carrotModifier;
+                } else if (this == Blocks.POTATOES) {
+                    modifier = level.spigotConfig.potatoModifier;
+                // Paper start - Fix Spigot growth modifiers
+                } else if (this == Blocks.TORCHFLOWER_CROP) {
+                    modifier = level.spigotConfig.torchFlowerModifier;
+                // Paper end - Fix Spigot growth modifiers
+                } else if (this == Blocks.WHEAT) {
+                    modifier = level.spigotConfig.wheatModifier;
+                }
+
+                modifier += modifier / 100 * growthPercentIncrease.getDouble(level.serverLevelData.settings.levelName);
+
+                if (random.nextFloat() < (modifier / (100.0F * (Math.floor((25.0F / growthSpeed) + 1))))) { // Spigot - SPIGOT-7159: Better modifier resolution
+                    // Spigot end
+                    org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockGrowEvent(level, pos, this.getStateForAge(age + 1), 2); // CraftBukkit
                 }
             }
         }
@@ -89,7 +_,7 @@
 
     public void growCrops(Level level, BlockPos pos, BlockState state) {
         int min = Math.min(this.getMaxAge(), this.getAge(state) + this.getBonemealAgeIncrease(level));
-        level.setBlock(pos, this.getStateForAge(min), Block.UPDATE_CLIENTS);
+        org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockGrowEvent(level, pos, this.getStateForAge(min), 2); // CraftBukkit
     }
 
     protected int getBonemealAgeIncrease(Level level) {
@@ -99,19 +_,20 @@
     protected static float getGrowthSpeed(Block block, BlockGetter level, BlockPos pos) {
         float f = 1.0F;
         BlockPos blockPos = pos.below();
+        boolean contains = growthPercentIncrease.containsKey(((ServerLevel)level).serverLevelData.settings.levelName);
 
         for (int i = -1; i <= 1; i++) {
             for (int i1 = -1; i1 <= 1; i1++) {
                 float f1 = 0.0F;
-                BlockState blockState = level.getBlockState(blockPos.offset(i, 0, i1));
-                if (blockState.is(Blocks.FARMLAND)) {
+                BlockState blockState = null;
+                if (contains || (blockState = level.getBlockState(blockPos.offset(i, 0, i1))).is(Blocks.FARMLAND)) {
                     f1 = 1.0F;
-                    if (blockState.getValue(FarmBlock.MOISTURE) > 0) {
+                    if (contains || blockState.getValue(FarmBlock.MOISTURE) > 0) {
                         f1 = 3.0F;
                     }
                 }
 
-                if (i != 0 || i1 != 0) {
+                if (!contains && (i != 0 || i1 != 0)) {
                     f1 /= 4.0F;
                 }
 
@@ -125,9 +_,9 @@
         BlockPos blockPos4 = pos.east();
         boolean flag = level.getBlockState(blockPos3).is(block) || level.getBlockState(blockPos4).is(block);
         boolean flag1 = level.getBlockState(blockPos1).is(block) || level.getBlockState(blockPos2).is(block);
-        if (flag && flag1) {
+        if (!contains && flag && flag1) {
             f /= 2.0F;
-        } else {
+        } else if(!contains) {
             boolean flag2 = level.getBlockState(blockPos3.north()).is(block)
                 || level.getBlockState(blockPos4.north()).is(block)
                 || level.getBlockState(blockPos4.south()).is(block)
@@ -137,6 +_,10 @@
             }
         }
 
+      if (contains) {
+        return f / 11;
+      }
+
         return f;
     }
 
@@ -146,12 +_,14 @@
     }
 
     protected static boolean hasSufficientLight(LevelReader level, BlockPos pos) {
-        return level.getRawBrightness(pos, 0) >= 8;
+        return level.getRawBrightness(pos, 0) >= 8 || (level instanceof ServerLevel serverLevel &&
+            org.bukkit.Bukkit.worldNamesWithAllowedGrowthWithLowBrightness.contains(serverLevel.getWorld().getName()));
     }
 
     @Override
     protected void entityInside(BlockState state, Level level, BlockPos pos, Entity entity, InsideBlockEffectApplier effectApplier, boolean pastEdges) {
-        if (level instanceof ServerLevel serverLevel && entity instanceof Ravager && serverLevel.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING)) {
+        if (!new io.papermc.paper.event.entity.EntityInsideBlockEvent(entity.getBukkitEntity(), org.bukkit.craftbukkit.block.CraftBlock.at(level, pos)).callEvent()) { return; } // Paper - Add EntityInsideBlockEvent
+        if (level instanceof ServerLevel serverLevel && entity instanceof Ravager && org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(entity, pos, Blocks.AIR.defaultBlockState(), !serverLevel.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING))) { // CraftBukkit
             serverLevel.destroyBlock(pos, true, entity);
         }
 
