--- a/net/minecraft/world/item/crafting/ShapelessRecipe.java
+++ b/net/minecraft/world/item/crafting/ShapelessRecipe.java
@@ -19,7 +_,7 @@
 public class ShapelessRecipe implements CraftingRecipe {
     final String group;
     final CraftingBookCategory category;
-    final ItemStack result;
+    public ItemStack result;
     final List<Ingredient> ingredients;
     @Nullable
     private PlacementInfo placementInfo;
@@ -30,6 +_,21 @@
         this.result = result;
         this.ingredients = ingredients;
     }
+
+    // CraftBukkit start
+    @Override
+    public org.bukkit.inventory.ShapelessRecipe toBukkitRecipe(org.bukkit.NamespacedKey id) {
+        org.bukkit.craftbukkit.inventory.CraftItemStack result = org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(this.result);
+        org.bukkit.craftbukkit.inventory.CraftShapelessRecipe recipe = new org.bukkit.craftbukkit.inventory.CraftShapelessRecipe(id, result, this);
+        recipe.setGroup(this.group);
+        recipe.setCategory(org.bukkit.craftbukkit.inventory.CraftRecipe.getCategory(this.category()));
+
+        for (Ingredient list : this.ingredients) {
+            recipe.addIngredient(org.bukkit.craftbukkit.inventory.CraftRecipe.toBukkit(list));
+        }
+        return recipe;
+    }
+    // CraftBukkit end
 
     @Override
     public RecipeSerializer<ShapelessRecipe> getSerializer() {
