--- a/net/minecraft/world/entity/projectile/ProjectileUtil.java
+++ b/net/minecraft/world/entity/projectile/ProjectileUtil.java
@@ -107,6 +_,8 @@
         return Math.max(0.0F, Math.min(0.3F, (entity.tickCount - 2) / 20.0F));
     }
 
+    public static final it.unimi.dsi.fastutil.ints.Int2DoubleOpenHashMap boundingBoxInflateMapForProjectileClip = new it.unimi.dsi.fastutil.ints.Int2DoubleOpenHashMap();
+
     @Nullable
     public static EntityHitResult getEntityHitResult(
         Level level, Entity source, Vec3 startVec, Vec3 endVec, AABB boundingBox, Predicate<Entity> filter, float margin
@@ -115,9 +_,16 @@
         Optional<Vec3> optional = Optional.empty();
         Entity entity = null;
 
+        final double v = boundingBoxInflateMapForProjectileClip.get(source.getId());
+        if (v != 0) {
+            boundingBox = boundingBox.inflate(v);
+        }
+
         for (Entity entity1 : level.getEntities(source, boundingBox, filter)) {
-            AABB aabb = entity1.getBoundingBox().inflate(margin);
-            Optional<Vec3> optional1 = aabb.clip(startVec, endVec);
+            final double v1 = boundingBoxInflateMapForProjectileClip.get(entity1.getId());
+            AABB aabb = entity1.getBoundingBox().inflate(margin + v + v1);
+            // без проверки на contains большой хитбокс не срабатывал
+            Optional<Vec3> optional1 = aabb.contains(startVec) ? Optional.of(startVec) : aabb.clip(startVec, endVec);
             if (optional1.isPresent()) {
                 double d1 = startVec.distanceToSqr(optional1.get());
                 if (d1 < d) {
