From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: loqqin <soubmatic@gmail.com>
Date: Sat, 11 Oct 2025 23:12:49 +0300
Subject: [PATCH] fixup! paper File Patches


diff --git a/net/minecraft/world/item/crafting/SingleItemRecipe.java b/net/minecraft/world/item/crafting/SingleItemRecipe.java
index 7c9d151f73ba055c4a3030fd443c942a9d89fabf..44c941c36cf2d81504d3fb0e3500363f85ce82c7 100644
--- a/net/minecraft/world/item/crafting/SingleItemRecipe.java
+++ b/net/minecraft/world/item/crafting/SingleItemRecipe.java
@@ -13,7 +13,7 @@ import net.minecraft.world.level.Level;
 
 public abstract class SingleItemRecipe implements Recipe<SingleRecipeInput> {
     private final Ingredient input;
-    public final ItemStack result;
+    public ItemStack result;
     private final String group;
     @Nullable
     private PlacementInfo placementInfo;
