From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: loqqin <soubmatic@gmail.com>
Date: Wed, 22 Oct 2025 18:32:51 +0300
Subject: [PATCH] fixup! paper File Patches


diff --git a/net/minecraft/world/entity/EntityType.java b/net/minecraft/world/entity/EntityType.java
index 28b511dfc50eb9335c2dec3bb31fd31f31925fc1..11fdb98eef4b70979013751206e2d54964023c19 100644
--- a/net/minecraft/world/entity/EntityType.java
+++ b/net/minecraft/world/entity/EntityType.java
@@ -1618,7 +1618,12 @@ public class EntityType<T extends Entity> implements FeatureElement, EntityTypeT
 
     private static Optional<Entity> loadStaticEntity(EntityType<?> entityType, ValueInput input, Level level, EntitySpawnReason spawnReason) {
         try {
+<<<<<<< HEAD
             return create(entityType, input, level, spawnReason);
+=======
+            return
+                create(entityType, valueInput, level, spawnReason);
+>>>>>>> fixup! paper File Patches
         } catch (RuntimeException var5) {
             LOGGER.warn("Exception loading entity: ", (Throwable)var5);
             return Optional.empty();
diff --git a/net/minecraft/world/entity/player/Player.java b/net/minecraft/world/entity/player/Player.java
index f9ad67668ca73d510e3165d68c1bef25e7069656..3327a18eba9ccd199d886c6df109029c9bfdc59c 100644
--- a/net/minecraft/world/entity/player/Player.java
+++ b/net/minecraft/world/entity/player/Player.java
@@ -1119,7 +1119,7 @@ public abstract class Player extends Avatar implements ContainerUser {
                                 // Paper end - Configurable sprint interruption on attack
                             }
 
-                            if (flag3) {
+                            if (flag3 && false) { // loqqin disable sweep
                                 float f5 = 1.0F + (float)this.getAttributeValue(Attributes.SWEEPING_DAMAGE_RATIO) * f;
 
                                 for (LivingEntity livingEntity2 : this.level()
