From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: aaa <121246396+loqqin@users.noreply.github.com>
Date: Tue, 18 Aug 2026 16:54:08 +0300
Subject: [PATCH] fixup! paper File Patches


diff --git a/io/papermc/paper/command/subcommands/FixLightCommand.java b/io/papermc/paper/command/subcommands/FixLightCommand.java
index c7303da057111baae585143979e7b1ffba0aad86..45d10a115ac25506b70816605558c7f18eaedfcc 100644
--- a/io/papermc/paper/command/subcommands/FixLightCommand.java
+++ b/io/papermc/paper/command/subcommands/FixLightCommand.java
@@ -35,7 +35,7 @@ public final class FixLightCommand implements PaperSubcommand {
         return true;
     }
 
-    private void doFixLight(final CommandSender sender, final String[] args) {
+    public static void doFixLight(final CommandSender sender, final String[] args) {
         if (!(sender instanceof Player)) {
             sender.sendMessage(text("Only players can use this command", RED));
             return;
@@ -64,10 +64,10 @@ public final class FixLightCommand implements PaperSubcommand {
         ServerPlayer handle = player.getHandle();
         ServerLevel world = (ServerLevel) handle.level();
         ThreadedLevelLightEngine lightengine = world.getChunkSource().getLightEngine();
-        this.starlightFixLight(handle, world, lightengine, radius, post);
+        starlightFixLight(handle, world, lightengine, radius, post);
     }
 
-    private void starlightFixLight(
+    public static void starlightFixLight(
         final ServerPlayer sender,
         final ServerLevel world,
         final ThreadedLevelLightEngine lightengine,
diff --git a/net/minecraft/server/ServerScoreboard.java b/net/minecraft/server/ServerScoreboard.java
index 7ffd683c42557052d7bfa963d87e68f4e8f62d6c..ffefbc1b7856120ab91659ebaa2337b6227118b8 100644
--- a/net/minecraft/server/ServerScoreboard.java
+++ b/net/minecraft/server/ServerScoreboard.java
@@ -294,6 +294,9 @@ public class ServerScoreboard extends Scoreboard {
     }
 
     private void updatePlayerWaypoint(final String player) {
+        if (true) { // loqqin fix lag on team set color
+            return;
+        }
         ServerPlayer serverPlayer = this.server.getPlayerList().getPlayerByName(player);
         if (serverPlayer != null) {
             serverPlayer.level().getWaypointManager().remakeConnections(serverPlayer);
diff --git a/net/minecraft/world/entity/player/Player.java b/net/minecraft/world/entity/player/Player.java
index 2a71a2fa949b3a7fb8ff2c44ec3a3d8636237142..54354faae9041b142ea82fa8158fa0d83eb71236 100644
--- a/net/minecraft/world/entity/player/Player.java
+++ b/net/minecraft/world/entity/player/Player.java
@@ -1257,6 +1257,7 @@ public abstract class Player extends Avatar implements ContainerUser {
     }
 
     private void doSweepAttack(final Entity entity, final float baseDamage, final DamageSource damageSource, final float attackStrengthScale) {
+        if (true) return; // loqqin disable sweep
         this.playServerSideSound(SoundEvents.PLAYER_ATTACK_SWEEP);
         if (this.level() instanceof ServerLevel serverLevel) {
             float var12 = 1.0F + (float)this.getAttributeValue(Attributes.SWEEPING_DAMAGE_RATIO) * baseDamage;
