From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: aaa <121246396+loqqin@users.noreply.github.com>
Date: Tue, 18 Aug 2026 16:54:35 +0300
Subject: [PATCH] loqqin


diff --git a/net/minecraft/world/item/crafting/ShapelessRecipe.java b/net/minecraft/world/item/crafting/ShapelessRecipe.java
index 53a818f14709d1978dbbc4aa7367be733a8ee47f..fb7eaa2a9c4b49b494e33ec7066b7850aa581842 100644
--- a/net/minecraft/world/item/crafting/ShapelessRecipe.java
+++ b/net/minecraft/world/item/crafting/ShapelessRecipe.java
@@ -36,8 +36,8 @@ public class ShapelessRecipe extends NormalCraftingRecipe {
         ShapelessRecipe::new
     );
     public static final RecipeSerializer<ShapelessRecipe> SERIALIZER = new RecipeSerializer<>(MAP_CODEC, STREAM_CODEC);
-    private final ItemStackTemplate result;
-    private final List<Ingredient> ingredients;
+    public final ItemStackTemplate result;
+    public final List<Ingredient> ingredients;
 
     public ShapelessRecipe(
         final Recipe.CommonInfo commonInfo, final CraftingRecipe.CraftingBookInfo bookInfo, final ItemStackTemplate result, final List<Ingredient> ingredients
