From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: loqqin <soubmatic@gmail.com>
Date: Thu, 5 Feb 2026 21:52:09 +0200
Subject: [PATCH] loqqin


diff --git a/net/minecraft/network/protocol/game/ClientboundMoveEntityPacket.java b/net/minecraft/network/protocol/game/ClientboundMoveEntityPacket.java
index 6d424f55380ac6daa1afc5fcf253048e095bb645..2b059be6cf3a6a2a5f56c883847e66d4cfa98eae 100644
--- a/net/minecraft/network/protocol/game/ClientboundMoveEntityPacket.java
+++ b/net/minecraft/network/protocol/game/ClientboundMoveEntityPacket.java
@@ -11,16 +11,16 @@ import net.minecraft.world.level.Level;
 
 public abstract class ClientboundMoveEntityPacket implements Packet<ClientGamePacketListener> {
     public final int entityId;
-    protected final short xa;
-    protected final short ya;
-    protected final short za;
-    protected final byte yRot;
-    protected final byte xRot;
-    protected final boolean onGround;
-    protected final boolean hasRot;
-    protected final boolean hasPos;
-
-    protected ClientboundMoveEntityPacket(int entityId, short xa, short ya, short za, byte yRot, byte xRot, boolean onGround, boolean hasRot, boolean hasPos) {
+    public final short xa;
+    public final short ya;
+    public final short za;
+    public final byte yRot;
+    public final byte xRot;
+    public final boolean onGround;
+    public final boolean hasRot;
+    public final boolean hasPos;
+
+    public ClientboundMoveEntityPacket(int entityId, short xa, short ya, short za, byte yRot, byte xRot, boolean onGround, boolean hasRot, boolean hasPos) {
         this.entityId = entityId;
         this.xa = xa;
         this.ya = ya;
