From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: aaa <121246396+loqqin@users.noreply.github.com>
Date: Tue, 18 Aug 2026 16:55:19 +0300
Subject: [PATCH] loqqin


diff --git a/net/minecraft/world/entity/ai/goal/FloatGoal.java b/net/minecraft/world/entity/ai/goal/FloatGoal.java
index 1b609edf8ad5ebb14b8a27384178eac4c0f1f58f..59cd69f512210066c0cfb5adc06a32f8849c2600 100644
--- a/net/minecraft/world/entity/ai/goal/FloatGoal.java
+++ b/net/minecraft/world/entity/ai/goal/FloatGoal.java
@@ -5,7 +5,7 @@ import net.minecraft.tags.FluidTags;
 import net.minecraft.world.entity.Mob;
 
 public class FloatGoal extends Goal {
-    private final Mob mob;
+    public final Mob mob;
 
     public FloatGoal(final Mob mob) {
         this.mob = mob;
