From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: aaa <121246396+loqqin@users.noreply.github.com>
Date: Tue, 18 Aug 2026 16:55:56 +0300
Subject: [PATCH] loqqin


diff --git a/net/minecraft/world/entity/projectile/FishingHook.java b/net/minecraft/world/entity/projectile/FishingHook.java
index c0bb1b62dfcdebbd34a8e8423fa36eb4924a55e5..78bc68e038206ae1fd71bf37f14416e48e511c71 100644
--- a/net/minecraft/world/entity/projectile/FishingHook.java
+++ b/net/minecraft/world/entity/projectile/FishingHook.java
@@ -357,6 +357,8 @@ public class FishingHook extends Projectile {
                     float particleZMovement = angleCos * 0.04F;
                     serverLevel.sendParticles(ParticleTypes.FISHING, fishX, fishY, fishZ, 0, particleZMovement, 0.01, -particleXMovement, 1.0);
                     serverLevel.sendParticles(ParticleTypes.FISHING, fishX, fishY, fishZ, 0, -particleZMovement, 0.01, particleXMovement, 1.0);
+                    org.bukkit.Bukkit.getPluginManager().callEvent(new org.bukkit.event.entity.FishingParticleEvent(((org.bukkit.entity.FishHook) this.getBukkitEntity()),
+                        new org.bukkit.Location(null, fishX, fishY, fishZ)));
                 }
             } else {
                 // CraftBukkit start
