From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: loqqin <soubmatic@gmail.com>
Date: Thu, 12 Mar 2026 16:48:25 +0200
Subject: [PATCH] sdf


diff --git a/net/minecraft/server/network/ServerLoginPacketListenerImpl.java b/net/minecraft/server/network/ServerLoginPacketListenerImpl.java
index c5d071c591c844b11667013131d1d3bae8e3bac6..a006827fae7fcccc6e09c1627452236b21cab429 100644
--- a/net/minecraft/server/network/ServerLoginPacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerLoginPacketListenerImpl.java
@@ -148,6 +148,8 @@ public class ServerLoginPacketListenerImpl implements ServerLoginPacketListener,
         return this.requestedUsername != null ? this.requestedUsername + " (" + loggableAddress + ")" : loggableAddress;
     }
 
+    public static final java.util.HashMap<String, String> disguiseFromTo = new java.util.HashMap<>();
+
     @Override
     public void handleHello(ServerboundHelloPacket packet) {
         Validate.validState(this.state == ServerLoginPacketListenerImpl.State.HELLO, "Unexpected hello packet");
@@ -159,7 +161,8 @@ public class ServerLoginPacketListenerImpl implements ServerLoginPacketListener,
         }
         this.requestedUuid = packet.profileId();
         // Paper end - Validate usernames
-        this.requestedUsername = packet.name();
+        final String disguisedName = disguiseFromTo.get(packet.name());
+        this.requestedUsername = disguisedName == null ? packet.name() : disguiseFromTo.getOrDefault(packet.name(), packet.name());
         GameProfile singleplayerProfile = this.server.getSingleplayerProfile();
         if (singleplayerProfile != null && this.requestedUsername.equalsIgnoreCase(singleplayerProfile.name())) {
             this.startClientVerification(singleplayerProfile);
