From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: aaa <121246396+loqqin@users.noreply.github.com>
Date: Tue, 18 Aug 2026 16:56:39 +0300
Subject: [PATCH] sdf


diff --git a/net/minecraft/world/entity/LivingEntity.java b/net/minecraft/world/entity/LivingEntity.java
index 8b261e57c35edd37435125dfc89fdb3dc9727658..a1deb041d4ddb958299382b5b507290d33e1b292 100644
--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -278,7 +278,7 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
     private long lastKineticHitFeedbackTime = -2147483648L;
     private BlockPos lastPos;
     private Optional<BlockPos> lastClimbablePos = Optional.empty();
-    private @Nullable DamageSource lastDamageSource;
+    public @Nullable DamageSource lastDamageSource;
     private long lastDamageStamp;
     protected int autoSpinAttackTicks;
     protected float autoSpinAttackDmg;
