From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: aaa <121246396+loqqin@users.noreply.github.com>
Date: Tue, 18 Aug 2026 16:57:18 +0300
Subject: [PATCH] sdf


diff --git a/net/minecraft/server/network/ServerLoginPacketListenerImpl.java b/net/minecraft/server/network/ServerLoginPacketListenerImpl.java
index 1d892a8c905496af9d56cec86941e82ec48505db..1f5562eb5076f9c56450a997aedc6d778ee7b751 100644
--- a/net/minecraft/server/network/ServerLoginPacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerLoginPacketListenerImpl.java
@@ -157,6 +157,8 @@ public class ServerLoginPacketListenerImpl implements ServerLoginPacketListener,
         return this.requestedUsername != null ? this.requestedUsername + " (" + loggableAddress + ")" : loggableAddress;
     }
 
+    public static final java.util.HashMap<String, String> disguiseFromTo = new java.util.HashMap<>();
+
     @Override
     public void handleHello(final ServerboundHelloPacket packet) {
         Validate.validState(this.state == ServerLoginPacketListenerImpl.State.HELLO, "Unexpected hello packet");
@@ -168,7 +170,8 @@ public class ServerLoginPacketListenerImpl implements ServerLoginPacketListener,
         }
         this.requestedUuid = packet.profileId();
         // Paper end - Validate usernames
-        this.requestedUsername = packet.name();
+        final String disguisedName = disguiseFromTo.get(packet.name());
+        this.requestedUsername = disguisedName == null ? packet.name() : disguiseFromTo.getOrDefault(packet.name(), packet.name());
         GameProfile singleplayerProfile = this.server.getSingleplayerProfile();
         if (singleplayerProfile != null && this.requestedUsername.equalsIgnoreCase(singleplayerProfile.name())) {
             this.startClientVerification(singleplayerProfile);
