From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: loqqin <soubmatic@gmail.com>
Date: Sat, 14 Mar 2026 14:09:05 +0200
Subject: [PATCH] sdf


diff --git a/net/minecraft/server/network/ServerLoginPacketListenerImpl.java b/net/minecraft/server/network/ServerLoginPacketListenerImpl.java
index b5d27cac93c6b2dc5640f0f9ffdadc25aa5c43fd..928a90cb15c6e938ae53af9ee5b10203099565e8 100644
--- a/net/minecraft/server/network/ServerLoginPacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerLoginPacketListenerImpl.java
@@ -170,8 +170,7 @@ public class ServerLoginPacketListenerImpl implements ServerLoginPacketListener,
         }
         this.requestedUuid = packet.profileId();
         // Paper end - Validate usernames
-        final String disguisedName = disguiseFromTo.get(packet.name());
-        this.requestedUsername = disguisedName == null ? packet.name() : disguiseFromTo.getOrDefault(packet.name().toLowerCase(java.util.Locale.ROOT), packet.name());
+        this.requestedUsername = disguiseFromTo.getOrDefault(packet.name().toLowerCase(java.util.Locale.ROOT), packet.name());
         GameProfile singleplayerProfile = this.server.getSingleplayerProfile();
         if (singleplayerProfile != null && this.requestedUsername.equalsIgnoreCase(singleplayerProfile.name())) {
             this.startClientVerification(singleplayerProfile);
