From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: aaa <121246396+loqqin@users.noreply.github.com>
Date: Tue, 18 Aug 2026 16:57:53 +0300
Subject: [PATCH] sdfsdf


diff --git a/net/minecraft/world/entity/projectile/ProjectileUtil.java b/net/minecraft/world/entity/projectile/ProjectileUtil.java
index 7a94272f5d773ead519f4ddee908722e3f4cbd7a..2fe2164d744ff38d3aacae50e3040f5e914ac879 100644
--- a/net/minecraft/world/entity/projectile/ProjectileUtil.java
+++ b/net/minecraft/world/entity/projectile/ProjectileUtil.java
@@ -26,6 +26,7 @@ import net.minecraft.world.phys.Vec3;
 import org.jspecify.annotations.Nullable;
 
 public final class ProjectileUtil {
+    public static final it.unimi.dsi.fastutil.ints.Int2DoubleOpenHashMap boundingBoxInflateMapForProjectileClip = new it.unimi.dsi.fastutil.ints.Int2DoubleOpenHashMap();
     public static final float DEFAULT_ENTITY_HIT_RESULT_MARGIN = 0.3F;
 
     public static HitResult getHitResultOnMoveVector(final Entity source, final Predicate<Entity> matching) {
@@ -168,7 +169,7 @@ public final class ProjectileUtil {
         Entity hitEntity = null;
 
         for (Entity entity : level.getEntities(source, targetSearchArea, matching)) {
-            AABB bb = entity.getBoundingBox().inflate(entityMargin);
+            AABB bb = entity.getBoundingBox().inflate(entityMargin + boundingBoxInflateMapForProjectileClip.getOrDefault(source.getId(), 0));
             Optional<Vec3> location = bb.clip(from, to);
             if (location.isPresent()) {
                 double dd = from.distanceToSqr(location.get());
