From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: aaa <121246396+loqqin@users.noreply.github.com>
Date: Fri, 8 May 2026 15:04:08 +0300
Subject: [PATCH] sdfsdf


diff --git a/net/minecraft/world/entity/projectile/Projectile.java b/net/minecraft/world/entity/projectile/Projectile.java
index cdb73cab0a64f51d65a7420fdac44455c6c77a48..c64896302099dc188fc594c84dc92b44a30df236 100644
--- a/net/minecraft/world/entity/projectile/Projectile.java
+++ b/net/minecraft/world/entity/projectile/Projectile.java
@@ -387,6 +387,8 @@ public abstract class Projectile extends Entity implements TraceableEntity {
     }
     // Paper end
 
+    public static final it.unimi.dsi.fastutil.ints.IntOpenHashSet canHitWithProjectileWhileCantSee = new it.unimi.dsi.fastutil.ints.IntOpenHashSet();
+
     protected boolean canHitEntity(Entity target) {
         if (!target.canBeHitByProjectile()) {
             return false;
@@ -396,7 +398,7 @@ public abstract class Projectile extends Entity implements TraceableEntity {
             if (owner instanceof net.minecraft.server.level.ServerPlayer) {
                 org.bukkit.entity.Entity collided = target.getBukkitEntity();
                 org.bukkit.entity.Player shooter = (org.bukkit.entity.Player) owner.getBukkitEntity();
-                if (!shooter.canSee(collided)) {
+                if (!shooter.canSee(collided) && !canHitWithProjectileWhileCantSee.contains(target.getId())) {
                     return false;
                 }
             }
