From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: aaa <121246396+loqqin@users.noreply.github.com>
Date: Tue, 18 Aug 2026 16:58:04 +0300
Subject: [PATCH] sdfsdf


diff --git a/net/minecraft/world/entity/monster/EnderMan.java b/net/minecraft/world/entity/monster/EnderMan.java
index 9b4ae8afda1f2202761cb49bc4f1181ce4eb55d5..5c60261fceec76451622498341aa8481212a7bd8 100644
--- a/net/minecraft/world/entity/monster/EnderMan.java
+++ b/net/minecraft/world/entity/monster/EnderMan.java
@@ -369,6 +369,8 @@ public class EnderMan extends Monster implements NeutralMob {
         return this.entityData.get(DATA_CARRY_STATE).orElse(null);
     }
 
+    public boolean projectileDamageAllowed = false;
+
     @Override
     public boolean hurtServer(final ServerLevel level, final DamageSource source, final float damage) {
         if (this.isInvulnerableTo(level, source)) {
@@ -376,7 +378,7 @@ public class EnderMan extends Monster implements NeutralMob {
         }
 
         AbstractThrownPotion thrownPotion = source.getDirectEntity() instanceof AbstractThrownPotion potion ? potion : null;
-        if (!source.is(DamageTypeTags.IS_PROJECTILE) && thrownPotion == null) { // Paper - EndermanEscapeEvent - diff on change - below logic relies on this path covering non-projectile damage.
+        if ((!source.is(DamageTypeTags.IS_PROJECTILE) || projectileDamageAllowed) && thrownPotion == null) { // Paper - EndermanEscapeEvent - diff on change - below logic relies on this path covering non-projectile damage.
             boolean result = super.hurtServer(level, source, damage);
             if (!(source.getEntity() instanceof LivingEntity) && this.random.nextInt(10) != 0) {
                 if (this.tryEscape(source.is(DamageTypeTags.IS_DROWNING) ? com.destroystokyo.paper.event.entity.EndermanEscapeEvent.Reason.DROWN : com.destroystokyo.paper.event.entity.EndermanEscapeEvent.Reason.CRITICAL_HIT)) { // Paper - EndermanEscapeEvent
