From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: aaa <121246396+loqqin@users.noreply.github.com>
Date: Tue, 18 Aug 2026 16:58:15 +0300
Subject: [PATCH] sdfsdf


diff --git a/net/minecraft/world/entity/projectile/ProjectileUtil.java b/net/minecraft/world/entity/projectile/ProjectileUtil.java
index 2fe2164d744ff38d3aacae50e3040f5e914ac879..afb8c9f79ce5c2569ee9e88ea4ff4f7e47107c47 100644
--- a/net/minecraft/world/entity/projectile/ProjectileUtil.java
+++ b/net/minecraft/world/entity/projectile/ProjectileUtil.java
@@ -168,9 +168,17 @@ public final class ProjectileUtil {
         Optional<Vec3> nearestLocation = Optional.empty();
         Entity hitEntity = null;
 
-        for (Entity entity : level.getEntities(source, targetSearchArea, matching)) {
-            AABB bb = entity.getBoundingBox().inflate(entityMargin + boundingBoxInflateMapForProjectileClip.getOrDefault(source.getId(), 0));
-            Optional<Vec3> location = bb.clip(from, to);
+        AABB searchArea = targetSearchArea;
+        final double v = boundingBoxInflateMapForProjectileClip.get(source.getId());
+        if (v != 0) {
+            searchArea = searchArea.inflate(v);
+        }
+
+        for (Entity entity : level.getEntities(source, searchArea, matching)) {
+            final double v1 = boundingBoxInflateMapForProjectileClip.get(entity.getId());
+            AABB bb = entity.getBoundingBox().inflate(entityMargin + v + v1);
+            // без проверки на contains большой хитбокс не срабатывал
+            Optional<Vec3> location = bb.contains(from) ? Optional.of(from) : bb.clip(from, to);
             if (location.isPresent()) {
                 double dd = from.distanceToSqr(location.get());
                 if (dd < nearest) {
