From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: aaa <121246396+loqqin@users.noreply.github.com>
Date: Thu, 25 Jun 2026 22:23:49 +0300
Subject: [PATCH] loqqin


diff --git a/net/minecraft/world/entity/Mob.java b/net/minecraft/world/entity/Mob.java
index f2a321d94ddc9247a3820d42f4a04290ea3ce6f6..0bf72f732ce9691cf0694a1f7b684dd7d8e7407e 100644
--- a/net/minecraft/world/entity/Mob.java
+++ b/net/minecraft/world/entity/Mob.java
@@ -131,7 +131,7 @@ public abstract class Mob extends LivingEntity implements EquipmentUser, Leashab
     protected MoveControl moveControl;
     protected JumpControl jumpControl;
     private final BodyRotationControl bodyRotationControl;
-    protected PathNavigation navigation;
+    public PathNavigation navigation;
     public GoalSelector goalSelector;
     public @Nullable net.minecraft.world.entity.ai.goal.FloatGoal goalFloat; // Paper - Allow nerfed mobs to jump and float
     public GoalSelector targetSelector;
