From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: aaa <121246396+loqqin@users.noreply.github.com>
Date: Tue, 18 Aug 2026 16:59:51 +0300
Subject: [PATCH] loqqin


diff --git a/net/minecraft/server/level/ChunkMap.java b/net/minecraft/server/level/ChunkMap.java
index 7e9361407042ddaabe3f66ee8957ae84b6ef2317..dc8a85c037d98962f0172c7aa1680d7ee8c60c60 100644
--- a/net/minecraft/server/level/ChunkMap.java
+++ b/net/minecraft/server/level/ChunkMap.java
@@ -405,7 +405,9 @@ public class ChunkMap extends SimpleRegionStorage implements ChunkHolder.PlayerP
         this.poiManager.tick(haveTime);
         profiler.popPush("chunk_unload");
         if (!this.level.noSave()) {
-            this.processUnloads(haveTime);
+            this.level.noSave = true;
+            System.out.println("process unloads " + this.level.serverLevelData.getLevelName());
+//            this.processUnloads(haveTime);
         }
 
         profiler.pop();
