From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: aaa <121246396+loqqin@users.noreply.github.com>
Date: Sat, 11 Jul 2026 17:25:13 +0300
Subject: [PATCH] loqqin


diff --git a/net/minecraft/server/level/ServerPlayerGameMode.java b/net/minecraft/server/level/ServerPlayerGameMode.java
index 44360b7963e1144396af05665a8ad3ce88251c6e..52c1232a5403deb0877219e8108f004a77bcbbbc 100644
--- a/net/minecraft/server/level/ServerPlayerGameMode.java
+++ b/net/minecraft/server/level/ServerPlayerGameMode.java
@@ -227,7 +227,7 @@ public class ServerPlayerGameMode {
 
                 this.destroyProgressStart = this.gameTicks;
                 float f = 1.0F;
-                BlockState blockState = this.level.getBlockState(pos);
+                BlockState blockState = FakeBlocks.function.apply(player, pos, level);
                 if (event.useInteractedBlock() != org.bukkit.event.Event.Result.DENY && !blockState.isAir()) { // Paper
                     EnchantmentHelper.onHitBlock(
                         this.level,
