From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: aaa <121246396+loqqin@users.noreply.github.com>
Date: Tue, 18 Aug 2026 17:01:13 +0300
Subject: [PATCH] loqqin


diff --git a/net/minecraft/server/level/ServerPlayerGameMode.java b/net/minecraft/server/level/ServerPlayerGameMode.java
index bc179fe6f8df2054e557f44c0461c192ba505412..bcbf4342694eed432f829a2e2a328575c61b99fc 100644
--- a/net/minecraft/server/level/ServerPlayerGameMode.java
+++ b/net/minecraft/server/level/ServerPlayerGameMode.java
@@ -229,7 +229,7 @@ public class ServerPlayerGameMode {
 
                 this.destroyProgressStart = this.gameTicks;
                 float progress = 1.0F;
-                BlockState blockState = this.level.getBlockState(pos);
+                BlockState blockState = net.FakeBlocks.function.apply(this.player, pos, this.level);
                 if (event.useInteractedBlock() != org.bukkit.event.Event.Result.DENY && !blockState.isAir()) { // Paper
                     EnchantmentHelper.onHitBlock(
                         this.level,
