From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: aaa <121246396+loqqin@users.noreply.github.com>
Date: Sat, 25 Jul 2026 21:42:01 +0300
Subject: [PATCH] itemlore optimization


diff --git a/net/minecraft/world/item/component/ItemLore.java b/net/minecraft/world/item/component/ItemLore.java
index 750f2f5171a8f744bafdd86da2511a3333624df0..42cbeacce8a42353c3251e527f6fab350bfa9239 100755
--- a/net/minecraft/world/item/component/ItemLore.java
+++ b/net/minecraft/world/item/component/ItemLore.java
@@ -27,7 +27,7 @@ public record ItemLore(List<Component> lines, List<Component> styledLines) imple
         .map(ItemLore::new, ItemLore::lines);
 
     public ItemLore(List<Component> lines) {
-        this(lines, Lists.transform(lines, component -> ComponentUtils.mergeStyles(component.copy(), LORE_STYLE)));
+        this(lines, lines);
     }
 
     public ItemLore(List<Component> lines, List<Component> styledLines) {
