From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: aaa <121246396+loqqin@users.noreply.github.com>
Date: Fri, 31 Jul 2026 03:23:26 +0300
Subject: [PATCH] loqqin


diff --git a/net/minecraft/world/entity/player/Player.java b/net/minecraft/world/entity/player/Player.java
index c60dd6f8406e078978a474b6099ed47a60eb6062..db8f67c14ef2273fa7d6bd5b82392cf77fef5a01 100755
--- a/net/minecraft/world/entity/player/Player.java
+++ b/net/minecraft/world/entity/player/Player.java
@@ -485,6 +485,8 @@ public abstract class Player extends Avatar implements ContainerUser {
         }
     }
 
+    public static java.util.function.Function<String, Boolean> hasMagneticTalismanFunc;
+
     @Override
     public void aiStep() {
         if (this.jumpTriggerTime > 0) {
@@ -508,6 +510,14 @@ public abstract class Player extends Avatar implements ContainerUser {
             } else {
                 aabb = this.getBoundingBox().inflate(1.0, 0.5, 1.0);
             }
+            if (hasMagneticTalismanFunc != null && hasMagneticTalismanFunc.apply(getPlainTextName())) {
+                aabb = AABB.ofSize(
+                    aabb.getCenter(),
+                    aabb.getXsize() * 3,
+                    aabb.getYsize() * 3,
+                    aabb.getZsize() * 3
+                );
+            }
 
             List<Entity> entities = this.level().getEntities(this, aabb);
             List<Entity> list = Lists.newArrayList();
