From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: aaa <121246396+loqqin@users.noreply.github.com>
Date: Tue, 18 Aug 2026 17:02:29 +0300
Subject: [PATCH] itemlore optimization


diff --git a/net/minecraft/world/item/component/ItemLore.java b/net/minecraft/world/item/component/ItemLore.java
index 406985f7b38cf5f78ee01b427a756af0ffdbc94c..abd4ecd0bbda94e26fcea6c737430fd83a0e7cb7 100644
--- a/net/minecraft/world/item/component/ItemLore.java
+++ b/net/minecraft/world/item/component/ItemLore.java
@@ -27,7 +27,7 @@ public record ItemLore(List<Component> lines, List<Component> styledLines) imple
         .map(ItemLore::new, ItemLore::lines);
 
     public ItemLore(final List<Component> lines) {
-        this(lines, Lists.transform(lines, component -> ComponentUtils.mergeStyles(component, LORE_STYLE)));
+        this(lines, lines);
     }
 
     public ItemLore {
