From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: aaa <121246396+loqqin@users.noreply.github.com>
Date: Wed, 5 Aug 2026 15:32:46 +0300
Subject: [PATCH] cactus farm disable


diff --git a/net/minecraft/world/level/block/CactusBlock.java b/net/minecraft/world/level/block/CactusBlock.java
index c69880ab05802a494bd96045ebeaa7b4b0d002d8..d782d42886c31dee8c91ed7f15f51f3f679ac5b5 100755
--- a/net/minecraft/world/level/block/CactusBlock.java
+++ b/net/minecraft/world/level/block/CactusBlock.java
@@ -56,7 +56,7 @@ public class CactusBlock extends Block {
     @Override
     protected void randomTick(BlockState state, ServerLevel level, BlockPos pos, RandomSource random) {
         BlockPos blockPos = pos.above();
-        if (level.isEmptyBlock(blockPos)) {
+        if (level.isEmptyBlock(blockPos) && canSurvive(state, level, blockPos)) {
             int i = 1;
             int ageValue = state.getValue(AGE);
 
