From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: aaa <121246396+loqqin@users.noreply.github.com>
Date: Tue, 18 Aug 2026 17:02:58 +0300
Subject: [PATCH] cactus farm disable


diff --git a/net/minecraft/world/level/block/CactusBlock.java b/net/minecraft/world/level/block/CactusBlock.java
index 017de22643a2ce14fc44bb0eb9b26c2d8a6f9381..27edd70466f603078c3b35a67bca9f701b8b2c18 100644
--- a/net/minecraft/world/level/block/CactusBlock.java
+++ b/net/minecraft/world/level/block/CactusBlock.java
@@ -55,7 +55,7 @@ public class CactusBlock extends Block {
     @Override
     protected void randomTick(final BlockState state, final ServerLevel level, final BlockPos pos, final RandomSource random) {
         BlockPos above = pos.above();
-        if (level.isEmptyBlock(above)) {
+        if (level.isEmptyBlock(above) && this.canSurvive(state, level, above)) {
             int height = 1;
             int age = state.getValue(AGE);
 
