From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: aaa <121246396+loqqin@users.noreply.github.com>
Date: Tue, 18 Aug 2026 17:03:06 +0300
Subject: [PATCH] fixup! paper File Patches


diff --git a/net/minecraft/network/chat/MutableComponent.java b/net/minecraft/network/chat/MutableComponent.java
index 96b25bac21eef2ed9d9cd41c012c74a808bc121f..8612f3d4cb90d51963562b9e7667790fc368a188 100644
--- a/net/minecraft/network/chat/MutableComponent.java
+++ b/net/minecraft/network/chat/MutableComponent.java
@@ -50,6 +50,9 @@ public final class MutableComponent implements Component {
     }
 
     public MutableComponent append(final Component component) {
+        if (component == this) {
+            throw new RuntimeException("sibling == this");
+        }
         this.siblings.add(component);
         return this;
     }
