From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: aaa <121246396+loqqin@users.noreply.github.com>
Date: Tue, 18 Aug 2026 17:03:24 +0300
Subject: [PATCH] lore test


diff --git a/net/minecraft/world/item/component/ItemLore.java b/net/minecraft/world/item/component/ItemLore.java
index 10b90245b98d980062765ea896026a9ab4d120ee..30c04e27f83437e32b97c3f28b387778c6ef6f24 100644
--- a/net/minecraft/world/item/component/ItemLore.java
+++ b/net/minecraft/world/item/component/ItemLore.java
@@ -27,8 +27,6 @@ public record ItemLore(List<Component> lines, List<Component> styledLines) imple
         .map(ItemLore::new, ItemLore::lines);
 
     public ItemLore(final List<? extends Component> lines) {
-        System.out.println(123 + " 123");
-        System.out.println("666");
         this((java.util.List<net.minecraft.network.chat.Component>) lines, (java.util.List<net.minecraft.network.chat.Component>) lines);
     }
 
