From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: aaa <121246396+loqqin@users.noreply.github.com>
Date: Tue, 18 Aug 2026 17:04:32 +0300
Subject: [PATCH] loqqin added entity.isAliveFast


diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index 9dceb53f3d969989d903eaf31675a7ae2567fc93..a42e7aa2221ee087b9be4d54d01994e50408e81f 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -3002,6 +3002,10 @@ public abstract class Entity
         return !this.isRemoved();
     }
 
+    public boolean isAliveFast() {
+        return !this.isRemoved();
+    }
+
     public boolean isInWall() {
         if (this.noPhysics) {
             return false;
diff --git a/net/minecraft/world/entity/LivingEntity.java b/net/minecraft/world/entity/LivingEntity.java
index 970e19fc6f8e0bccf94022fba90713cbaab06266..053d8f6fa4164ed5324923245d1178f283907bd7 100644
--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -2190,6 +2190,11 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
         return !this.isRemoved() && this.getHealth() > 0.0F && !this.dead; // Paper - Check this.dead
     }
 
+    @Override
+    public boolean isAliveFast() {
+        return !this.isRemoved() && !this.dead; // Paper - Check this.dead
+    }
+
     public boolean isLookingAtMe(
         final LivingEntity target,
         final double coneSize,
