From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: aaa <121246396+loqqin@users.noreply.github.com>
Date: Tue, 18 Aug 2026 17:04:52 +0300
Subject: [PATCH] loqqin optimized ItemLore.equals


diff --git a/net/minecraft/world/item/component/ItemLore.java b/net/minecraft/world/item/component/ItemLore.java
index 30c04e27f83437e32b97c3f28b387778c6ef6f24..f98ea115b62cfffd55fcb97959d6c2fae9e8d274 100644
--- a/net/minecraft/world/item/component/ItemLore.java
+++ b/net/minecraft/world/item/component/ItemLore.java
@@ -46,4 +46,31 @@ public record ItemLore(List<Component> lines, List<Component> styledLines) imple
     ) {
         this.styledLines.forEach(consumer);
     }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (!(o instanceof ItemLore other)) return false;
+
+
+        if (this.lines.size() != other.lines.size()) return false;
+
+
+        if (!this.lines.equals(other.lines)) return false;
+
+
+        if (this.lines == this.styledLines && other.lines == other.styledLines) {
+            return true;
+        }
+
+        return this.styledLines.equals(other.styledLines);
+    }
+
+    @Override
+    public int hashCode() {
+        if (this.lines == this.styledLines) {
+            return this.lines.hashCode();
+        }
+        return 31 * this.lines.hashCode() + this.styledLines.hashCode();
+    }
 }
