--- /dev/null
+++ b/ca/spottedleaf/moonrise/patches/block_counting/BlockCountingBitStorage.java
@@ -1,0 +_,10 @@
+package ca.spottedleaf.moonrise.patches.block_counting;
+
+import it.unimi.dsi.fastutil.ints.Int2ObjectOpenHashMap;
+import it.unimi.dsi.fastutil.shorts.ShortArrayList;
+
+public interface BlockCountingBitStorage {
+
+    public Int2ObjectOpenHashMap<ShortArrayList> moonrise$countEntries();
+
+}
