--- a/net/minecraft/world/entity/ai/goal/FloatGoal.java
+++ b/net/minecraft/world/entity/ai/goal/FloatGoal.java
@@ -5,10 +_,11 @@
 import net.minecraft.world.entity.Mob;
 
 public class FloatGoal extends Goal {
-    private final Mob mob;
+    public final Mob mob;
 
     public FloatGoal(Mob mob) {
         this.mob = mob;
+        if (mob.level().paperConfig().entities.behavior.spawnerNerfedMobsShouldJump) mob.goalFloat = this; // Paper - Allow nerfed mobs to jump and float
         this.setFlags(EnumSet.of(Goal.Flag.JUMP));
         mob.getNavigation().setCanFloat(true);
     }
