--- a/net/minecraft/SharedConstants.java
+++ b/net/minecraft/SharedConstants.java
@@ -78,7 +_,7 @@
     public static final boolean DEBUG_POI = debugFlag("POI");
     public static final boolean DEBUG_BEES = debugFlag("BEES");
     public static final boolean DEBUG_RAIDS = debugFlag("RAIDS");
-    public static final boolean DEBUG_BLOCK_BREAK = debugFlag("BLOCK_BREAK");
+    public static boolean DEBUG_BLOCK_BREAK = debugFlag("BLOCK_BREAK");
     public static final boolean DEBUG_MONITOR_TICK_TIMES = debugFlag("MONITOR_TICK_TIMES");
     public static final boolean DEBUG_KEEP_JIGSAW_BLOCKS_DURING_STRUCTURE_GEN = debugFlag("KEEP_JIGSAW_BLOCKS_DURING_STRUCTURE_GEN");
     public static final boolean DEBUG_DONT_SAVE_WORLD = debugFlag("DONT_SAVE_WORLD");
