--- a/net/minecraft/util/parsing/packrat/Scope.java
+++ b/net/minecraft/util/parsing/packrat/Scope.java
@@ -281,4 +_,32 @@
 
         return true;
     }
+
+    // Paper start - track depth
+    private int depth;
+    @SuppressWarnings({"unchecked","rawtypes"})
+    public static <S> Term<S> increaseDepth() {
+        return new Term<S>() {
+            @Override
+            public boolean parse(final ParseState<S> parseState, final Scope scope, final Control control) {
+                if (++scope.depth > 512) {
+                    parseState.errorCollector().store(parseState.mark(), new IllegalStateException("Too deep"));
+                    return false;
+                }
+                return true;
+            }
+        };
+    }
+
+    @SuppressWarnings({"unchecked","rawtypes"})
+    public static <S> Term<S> decreaseDepth() {
+        return new net.minecraft.util.parsing.packrat.Term<S>() {
+            @Override
+            public boolean parse(final ParseState<S> parseState, final Scope scope, final Control control) {
+                scope.depth--;
+                return true;
+            }
+        };
+    }
+    // Paper end - track depth
 }
