--- a/net/minecraft/world/item/component/ItemLore.java
+++ b/net/minecraft/world/item/component/ItemLore.java
@@ -27,7 +_,7 @@
         .map(ItemLore::new, ItemLore::lines);
 
     public ItemLore(List<Component> lines) {
-        this(lines, Lists.transform(lines, component -> ComponentUtils.mergeStyles(component.copy(), LORE_STYLE)));
+        this(lines, lines);
     }
 
     public ItemLore(List<Component> lines, List<Component> styledLines) {
