--- a/net/minecraft/world/entity/monster/MagmaCube.java
+++ b/net/minecraft/world/entity/monster/MagmaCube.java
@@ -35,9 +_,11 @@
     }
 
     @Override
-    public void setSize(int size, boolean resetHealth) {
-        super.setSize(size, resetHealth);
-        this.getAttribute(Attributes.ARMOR).setBaseValue(size * 3);
+    public void setSize(int size, boolean resetHealth, boolean ignoreAttributes) {
+        super.setSize(size, resetHealth, ignoreAttributes);
+        if (!ignoreAttributes) {
+            this.getAttribute(Attributes.ARMOR).setBaseValue(size * 3);
+        }
     }
 
     @Override
